File: server/ordersLogic.js
/** @typedef {import('better-sqlite3').Database} Db */

export function marketItems(items = []) {
  return items.filter(it => it.source === 'market' || (!it.restId && it.source !== 'restaurant'))
}

export function restItems(items = [], restId) {
  return items.filter(it => {
    if (it.source === 'restaurant' || it.restId) {
      return restId == null || String(it.restId) === String(restId)
    }
    return false
  })
}

export function inferType(order) {
  const items = order.items || []
  const hasM = marketItems(items).length > 0
  const hasR = restItems(items).length > 0
  if (hasM && hasR) return 'mixed'
  if (hasR) return 'restaurant'
  if (hasM) return 'market'
  return order.type || 'market'
}

export function getMarketStatus(order) {
  if (order.marketStatus) return order.marketStatus
  const items = order.items || []
  if (!marketItems(items).length) return 'done'
  const mkt = marketItems(items)
  if (mkt.length > 0 && mkt.every(it => it.done)) return 'done'
  if (order.status === 'assembler_done') return 'done'
  if (inferType(order) === 'market' && ['courier_picked', 'delivering', 'delivered'].includes(order.status)) {
    return 'done'
  }
  if (order.status === 'assembling') return 'assembling'
  return 'new'
}

export function getRestPartStatus(order, restId) {
  const parts = order.restParts || {}
  const key = String(restId)
  if (parts[key]) return parts[key]
  if (!restItems(order.items || [], restId).length) return 'done'
  const mixed = inferType(order) === 'mixed'
  const hasMarket = marketItems(order.items || []).length > 0
  if (!mixed || !hasMarket) {
    const st = order.status
    if (['ready', 'assembler_done', 'courier_picked', 'delivering', 'delivered'].includes(st)) return 'done'
    if (st === 'cooking') return 'cooking'
    return 'new'
  }
  return 'new'
}

export function allPartsDone(order) {
  const items = order.items || []
  if (marketItems(items).length && getMarketStatus(order) !== 'done') return false
  const ids = new Set()
  if (order.restId) ids.add(String(order.restId))
  for (const rid of order.restIds || []) ids.add(String(rid))
  for (const it of restItems(items)) if (it.restId) ids.add(String(it.restId))
  for (const rid of ids) {
    if (getRestPartStatus(order, rid) !== 'done') return false
  }
  return true
}

export function anyPartDone(order) {
  const items = order.items || []
  if (marketItems(items).length && getMarketStatus(order) === 'done') return true
  const ids = new Set()
  if (order.restId) ids.add(String(order.restId))
  for (const rid of order.restIds || []) ids.add(String(rid))
  for (const it of restItems(items)) if (it.restId) ids.add(String(it.restId))
  for (const rid of ids) {
    if (getRestPartStatus(order, rid) === 'done') return true
  }
  return false
}

export function isAssemblerOrder(order) {
  const t = inferType(order)
  if (t === 'mixed') return ['new', 'assembling'].includes(getMarketStatus(order))
  return t === 'market' && ['new', 'assembling'].includes(order.status)
}

export function isCourierReady(order) {
  if (['delivered', 'cancelled', 'courier_picked', 'delivering'].includes(order.status)) return false
  const t = inferType(order)
  if (t === 'mixed') return anyPartDone(order) && ['ready', 'assembler_done'].includes(order.status)
  if (t === 'market') return order.status === 'assembler_done'
  if (t === 'restaurant') return order.status === 'ready' || allPartsDone(order)
  return false
}

export function isCourierSync(order) {
  if (['delivered', 'cancelled'].includes(order.status)) return false
  return isCourierReady(order) || ['courier_picked', 'delivering'].includes(order.status)
}

export function isCourierMapOrder(order) {
  if (['delivered', 'cancelled', 'courier_picked', 'delivering'].includes(order.status)) return false
  return ['new', 'assembling', 'cooking', 'ready', 'assembler_done'].includes(order.status)
}

export function isCourierMapSync(order) {
  if (['delivered', 'cancelled'].includes(order.status)) return false
  return isCourierMapOrder(order) || ['courier_picked', 'delivering'].includes(order.status)
}

export function syncMixedStatus(order) {
  if (inferType(order) !== 'mixed') return order
  if (['courier_picked', 'delivering', 'delivered', 'cancelled'].includes(order.status)) return order
  if (allPartsDone(order)) order.status = 'assembler_done'
  else if (anyPartDone(order)) order.status = 'ready'
  else if (getMarketStatus(order) === 'assembling') order.status = 'assembling'
  else order.status = 'new'
  return order
}

function collectRestIds(order) {
  const ids = new Set()
  if (order.restId) ids.add(String(order.restId))
  for (const rid of order.restIds || []) ids.add(String(rid))
  for (const it of restItems(order.items || [])) {
    if (it.restId) ids.add(String(it.restId))
  }
  return [...ids]
}

/** Админка: принудительно ставит статус и согласует этапы заказа */
export function applyAdminStatusFields(order, newStatus) {
  const t = inferType(order)
  order.status = newStatus

  if (!['courier_picked', 'delivering', 'delivered'].includes(newStatus)) {
    order.courier = null
    order.courierRoute = null
    order.pickedUpIds = []
  }
  if (newStatus === 'delivered' && !order.deliveredAt) {
    order.deliveredAt = new Date().toLocaleTimeString('ru-RU', { hour: '2-digit', minute: '2-digit' })
  }

  if (newStatus === 'new') {
    order.assembler = null
  }

  const items = order.items || []
  const restIds = collectRestIds(order)
  const hasM = marketItems(items).length > 0

  if (t === 'mixed') {
    if (['new', 'cancelled'].includes(newStatus)) {
      if (hasM) order.marketStatus = 'new'
      if (restIds.length) order.restParts = Object.fromEntries(restIds.map(id => [id, 'new']))
      order.items = items.map(it => ({ ...it, done: false }))
    } else if (newStatus === 'assembling') {
      if (hasM) order.marketStatus = 'assembling'
      if (restIds.length) order.restParts = Object.fromEntries(restIds.map(id => [id, 'new']))
      order.items = items.map(it => ({ ...it, done: false }))
    } else if (newStatus === 'cooking') {
      if (hasM) order.marketStatus = 'new'
      if (restIds.length) order.restParts = Object.fromEntries(restIds.map(id => [id, 'cooking']))
      order.items = items.map(it => ({ ...it, done: false }))
    } else if (newStatus === 'ready') {
      if (hasM) order.marketStatus = 'done'
      if (restIds.length) order.restParts = Object.fromEntries(restIds.map(id => [id, 'new']))
      order.items = items.map(it => (marketItems([it]).length ? { ...it, done: true } : { ...it, done: false }))
    } else if (newStatus === 'assembler_done') {
      if (hasM) order.marketStatus = 'done'
      if (restIds.length) order.restParts = Object.fromEntries(restIds.map(id => [id, 'done']))
      order.items = items.map(it => ({ ...it, done: true }))
    } else if (['courier_picked', 'delivering', 'delivered'].includes(newStatus)) {
      if (hasM) order.marketStatus = 'done'
      if (restIds.length) order.restParts = Object.fromEntries(restIds.map(id => [id, 'done']))
      order.items = items.map(it => ({ ...it, done: true }))
    }
  } else if (t === 'market') {
    if (['new', 'cancelled', 'assembling', 'cooking'].includes(newStatus)) {
      order.items = items.map(it => ({ ...it, done: false }))
    } else {
      order.items = items.map(it => ({ ...it, done: true }))
    }
  }

  return order
}

export function applyStatusPatch(order, body) {
  if (body.adminOverride === true && body.status) {
    applyAdminStatusFields(order, body.status)
    if ('courier' in body) order.courier = body.courier ?? null
    if ('assembler' in body) order.assembler = body.assembler ?? null
    if (body.deliveredAt != null) order.deliveredAt = body.deliveredAt
    if (body.deliveryFee != null) order.deliveryFee = body.deliveryFee
    if (body.deliveryFeeLocked != null) order.deliveryFeeLocked = body.deliveryFeeLocked
    return order
  }

  const adminOverride = body.adminOverride === true

  if (body.marketStatus != null) order.marketStatus = body.marketStatus
  if (body.restParts != null) {
    order.restParts = adminOverride
      ? body.restParts
      : { ...(order.restParts || {}), ...body.restParts }
  }
  if (body.pickedUpIds != null) order.pickedUpIds = body.pickedUpIds
  if ('courierRoute' in body) order.courierRoute = body.courierRoute ?? null
  if ('courier' in body) order.courier = body.courier ?? null
  if ('assembler' in body) order.assembler = body.assembler ?? null
  if (body.deliveredAt != null) order.deliveredAt = body.deliveredAt
  if (body.deliveryFee != null) order.deliveryFee = body.deliveryFee
  if (body.deliveryFeeLocked != null) order.deliveryFeeLocked = body.deliveryFeeLocked
  if (body.items != null) order.items = body.items
  if ('courierAtClient' in body) order.courierAtClient = !!body.courierAtClient

  if (body.status) {
    order.status = body.status
    if (body.status === 'delivered') order.courierAtClient = false
  } else if (!adminOverride && inferType(order) === 'mixed') {
    syncMixedStatus(order)
  }
  return order
}

File: backend/orders_logic.py
"""Логика заказов (смешанные / сборщик / курьер / ресторан)."""


def _market_items(items: list) -> list:
    return [it for it in items if it.get("source") == "market" or (not it.get("restId") and it.get("source") != "restaurant")]


def _rest_items(items: list, rest_id: str | None = None) -> list:
    out = []
    for it in items:
        if it.get("source") == "restaurant" or it.get("restId"):
            if rest_id is None or str(it.get("restId")) == str(rest_id):
                out.append(it)
    return out


def infer_type(order: dict) -> str:
    items = order.get("items") or []
    has_market = len(_market_items(items)) > 0
    has_rest = len(_rest_items(items)) > 0
    if has_market and has_rest:
        return "mixed"
    if has_rest:
        return "restaurant"
    return order.get("type") or "market"


def get_market_status(order: dict) -> str:
    if order.get("marketStatus"):
        return order["marketStatus"]
    items = order.get("items") or []
    if not _market_items(items):
        return "done"
    st = order.get("status", "new")
    if st == "assembler_done":
        return "done"
    if st == "assembling":
        return "assembling"
    return "new"


def get_rest_part_status(order: dict, rest_id: str) -> str:
    parts = order.get("restParts") or {}
    if rest_id in parts:
        return parts[rest_id]
    if not _rest_items(order.get("items") or [], rest_id):
        return "done"
    if infer_type(order) == "mixed":
        return "new"
    st = order.get("status", "new")
    if st in ("ready", "assembler_done", "courier_picked", "delivering", "delivered"):
        return "done"
    if st == "cooking":
        return "cooking"
    return "new"


def all_parts_done(order: dict) -> bool:
    items = order.get("items") or []
    if _market_items(items) and get_market_status(order) != "done":
        return False
    rest_ids = set()
    if order.get("restId"):
        rest_ids.add(str(order["restId"]))
    for rid in order.get("restIds") or []:
        rest_ids.add(str(rid))
    for it in _rest_items(items):
        if it.get("restId"):
            rest_ids.add(str(it["restId"]))
    for rid in rest_ids:
        if get_rest_part_status(order, rid) != "done":
            return False
    return True


def is_assembler_order(order: dict) -> bool:
    otype = infer_type(order)
    if otype == "mixed":
        ms = get_market_status(order)
        return ms in ("new", "assembling")
    return otype == "market" and order.get("status") in ("new", "assembling")


def is_courier_ready(order: dict) -> bool:
    st = order.get("status")
    if st in ("delivered", "cancelled"):
        return False
    otype = infer_type(order)
    if otype == "mixed":
        return all_parts_done(order) and st == "assembler_done"
    if otype == "market":
        return st == "assembler_done"
    if otype == "restaurant":
        return st == "ready"
    return False


def is_courier_sync(order: dict) -> bool:
    st = order.get("status")
    if st in ("delivered", "cancelled"):
        return False
    return is_courier_ready(order) or st in ("courier_picked", "delivering")


def sync_mixed_status(order: dict) -> dict:
    otype = infer_type(order)
    if otype != "mixed":
        return order
    if all_parts_done(order):
        order["status"] = "assembler_done"
    elif get_market_status(order) == "assembling":
        order["status"] = "assembling"
    else:
        order["status"] = "new"
    return order


def _collect_rest_ids(order: dict) -> list[str]:
    ids = set()
    if order.get("restId"):
        ids.add(str(order["restId"]))
    for rid in order.get("restIds") or []:
        ids.add(str(rid))
    for it in _rest_items(order.get("items") or []):
        if it.get("restId"):
            ids.add(str(it["restId"]))
    return list(ids)


def apply_admin_status_fields(order: dict, new_status: str) -> dict:
    otype = infer_type(order)
    order["status"] = new_status

    if new_status not in ("courier_picked", "delivering", "delivered"):
        order["courier"] = None
        order["courierRoute"] = None
        order["pickedUpIds"] = []

    if new_status == "new":
        order["assembler"] = None

    items = order.get("items") or []
    rest_ids = _collect_rest_ids(order)
    has_m = len(_market_items(items)) > 0

    if otype == "mixed":
        if new_status in ("new", "cancelled"):
            if has_m:
                order["marketStatus"] = "new"
            if rest_ids:
                order["restParts"] = {rid: "new" for rid in rest_ids}
            order["items"] = [{**it, "done": False} for it in items]
        elif new_status == "assembling":
            if has_m:
                order["marketStatus"] = "assembling"
            if rest_ids:
                order["restParts"] = {rid: "new" for rid in rest_ids}
            order["items"] = [{**it, "done": False} for it in items]
        elif new_status == "cooking":
            if has_m:
                order["marketStatus"] = "new"
            if rest_ids:
                order["restParts"] = {rid: "cooking" for rid in rest_ids}
            order["items"] = [{**it, "done": False} for it in items]
        elif new_status == "ready":
            if has_m:
                order["marketStatus"] = "done"
            if rest_ids:
                order["restParts"] = {rid: "new" for rid in rest_ids}
            order["items"] = [
                {**it, "done": True} if _market_items([it]) else {**it, "done": False}
                for it in items
            ]
        elif new_status == "assembler_done":
            if has_m:
                order["marketStatus"] = "done"
            if rest_ids:
                order["restParts"] = {rid: "done" for rid in rest_ids}
            order["items"] = [{**it, "done": True} for it in items]
        elif new_status in ("courier_picked", "delivering", "delivered"):
            if has_m:
                order["marketStatus"] = "done"
            if rest_ids:
                order["restParts"] = {rid: "done" for rid in rest_ids}
            order["items"] = [{**it, "done": True} for it in items]
    elif otype == "market":
        if new_status in ("new", "cancelled", "assembling", "cooking"):
            order["items"] = [{**it, "done": False} for it in items]
        else:
            order["items"] = [{**it, "done": True} for it in items]

    return order


def apply_status_patch(order: dict, body: dict) -> dict:
    if body.get("adminOverride") and body.get("status"):
        order = apply_admin_status_fields(order, body["status"])
        if "courier" in body:
            order["courier"] = body.get("courier")
        if "assembler" in body:
            order["assembler"] = body.get("assembler")
        if body.get("deliveredAt") is not None:
            order["deliveredAt"] = body.get("deliveredAt")
        return order

    status = body.get("status")
    if body.get("marketStatus") is not None:
        order["marketStatus"] = body["marketStatus"]
    if body.get("restParts") is not None:
        order["restParts"] = {**(order.get("restParts") or {}), **body["restParts"]}
    if body.get("items") is not None:
        order["items"] = body["items"]
    if "courier" in body:
        order["courier"] = body.get("courier")
    if "assembler" in body:
        order["assembler"] = body.get("assembler")
    if body.get("pickedUpIds") is not None:
        order["pickedUpIds"] = body["pickedUpIds"]

    otype = infer_type(order)
    if status:
        order["status"] = status
    elif otype == "mixed":
        order = sync_mixed_status(order)
    return order
